# Practical 6: Sentiment Analysis
#### Robert Bagheri
<img src="img/uu_logo.png" alt="logo" align="right" title="UU" width="50" height="20" />

#### Applied Text Mining - Utrecht Summer School

In this practical, we will apply both dictionary- and deep learning-based sentiment analysis approaches on the IMDB sentiment classification task.

We are going to use the following libraries. Take care to have them installed!

In [1]:
%%capture
%pip install -q vaderSentiment
%pip install -q --upgrade --force-reinstall --no-deps tensorflow-datasets==4.9.10 protobuf==6.31.1
%pip install -q transformers==4.57.6
# THIS IS A TEMPORARY FIX FOR CURRENTLY BROKEN COLAB ENV

In [2]:
%%capture
import time

import numpy as np
import pandas as pd

import tensorflow as tf
import tensorflow_datasets as tfds
import tensorflow_hub as hub
import tf_keras

from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from sklearn import metrics

# The warning about TensorFlow's Protobuf-generated code mismatch is benign and can be ignored

/usr/local/lib/python3.12/dist-packages/google/protobuf/runtime_version.py:98: UserWarning: Protobuf gencode version 5.28.3 is exactly one major version older than the runtime version 6.31.1 at tensorflow/core/framework/attr_value.proto. Please update the gencode to avoid compatibility violations in the next runtime release.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/google/protobuf/runtime_version.py:98: UserWarning: Protobuf gencode version 5.28.3 is exactly one major version older than the runtime version 6.31.1 at tensorflow/core/framework/tensor.proto. Please update the gencode to avoid compatibility violations in the next runtime release.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/google/protobuf/runtime_version.py:98: UserWarning: Protobuf gencode version 5.28.3 is exactly one major version older than the runtime version 6.31.1 at tensorflow/core/framework/resource_handle.proto. Please update the gencode to avoid compatibility violations in the next r

### Let's get started!

Here we are going to classify movie reviews as positive or negative using the text of the review. We will use the IMDB dataset that contains the text of 50,000 movie reviews from the Internet Movie Database ([IMDb](https://www.imdb.com/)). These are split into 25,000 reviews for training and 25,000 reviews for testing. The training and test sets are balanced, meaning they contain an equal number of positive and negative reviews.


1\. **The IMDB dataset is available on TensorFlow datasets. Use the following code to download the IMDB dataset.**

In [3]:
# Split the training set into 60% and 40% to end up with 15,000 examples
# for training, 10,000 examples for validation and 25,000 examples for testing.
train_data, validation_data, test_data = tfds.load(
    name="imdb_reviews",
    split=('train[:60%]', 'train[60%:]', 'test'),
    as_supervised=True)

Dl Completed...: 0 url [00:00, ? url/s]

Dl Size...: 0 MiB [00:00, ? MiB/s]

Generating splits...:   0%|          | 0/3 [00:00<?, ? splits/s]

Generating train examples...: 0 examples [00:00, ? examples/s]

Shuffling /root/tensorflow_datasets/imdb_reviews/plain_text/incomplete.P6G81W_1.0.0/imdb_reviews-train.tfrecor…

Generating test examples...: 0 examples [00:00, ? examples/s]

Shuffling /root/tensorflow_datasets/imdb_reviews/plain_text/incomplete.P6G81W_1.0.0/imdb_reviews-test.tfrecord…

Generating unsupervised examples...: 0 examples [00:00, ? examples/s]

Shuffling /root/tensorflow_datasets/imdb_reviews/plain_text/incomplete.P6G81W_1.0.0/imdb_reviews-unsupervised.…

Dataset imdb_reviews downloaded and prepared to /root/tensorflow_datasets/imdb_reviews/plain_text/1.0.0. Subsequent calls will reuse this data.


2\. **Use the following code to explore the data and print the first 4 examples.**

In [4]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(4)))
train_examples_batch

<tf.Tensor: shape=(4,), dtype=string, numpy=
array([b"This was an absolutely terrible movie. Don't be lured in by Christopher Walken or Michael Ironside. Both are great actors, but this must simply be their worst role in history. Even their great acting could not redeem this movie's ridiculous storyline. This movie is an early nineties US propaganda piece. The most pathetic scenes were those when the Columbian rebels were making their cases for revolutions. Maria Conchita Alonso appeared phony, and her pseudo-love affair with Walken was nothing but a pathetic emotional plug in a movie that was devoid of any real meaning. I am disappointed that there are movies like this, ruining actor's like Christopher Walken's good name. I could barely sit through it.",
       b'I have been known to fall asleep during films, but this is usually due to a combination of things including, really tired, being warm and comfortable on the sette and having just eaten a lot. However on this occasion I fell a

In [5]:
train_labels_batch

<tf.Tensor: shape=(4,), dtype=int64, numpy=array([0, 0, 0, 1])>

The label is an integer value of either 0 or 1, where 0 is a negative review, and 1 is a positive review.

You can convert the TensorFlow dataset to a Pandas DataFrame for easier inspection.

In [6]:
train_df = pd.DataFrame([
    {
        "text": text.numpy().decode("utf-8"),
        "label": int(label.numpy())
    }
    for text, label in train_data
])

test_df = pd.DataFrame([
    {
        "text": text.numpy().decode("utf-8"),
        "label": int(label.numpy())
    }
    for text, label in test_data
])

# Lexicon-based sentiment analysis

Vader (Valence Aware Dictionary and sEntiment Reasoner) is a lexicon and rule-based sentiment analysis tool that is specifically attuned to sentiments expressed in social media, and works well on texts from other domains.

The [VADER lexicon](https://www.kaggle.com/datasets/nltkdata/vader-lexicon) is empirically validated by multiple independent human judges. VADER incorporates a "gold-standard" sentiment lexicon that is especially attuned to microblog-like contexts.

It has some  advantages:
- Unsupervised
- Fast and deployable
- Reasonable performance even without preprocessing

However, there are some disadvantages:
- It is a rule-based approach, meaning it utilizes a list of predefined polarity scores for each word
- It cannot exceed beyond a certain performance compared to state-of-the-art NLP approaches

3\. **Create a Vader analyzer using the `SentimentIntensityAnalyzer` function, and look at the polarity scores of some example sentences.**

In [7]:
vader_example_sentence = "you cannot be negative"

analyzer = SentimentIntensityAnalyzer()
example_scores = analyzer.polarity_scores(vader_example_sentence)

print("Sentence:", vader_example_sentence)
print("VADER output:", example_scores)

Sentence: you cannot be negative
VADER output: {'neg': 0.0, 'neu': 0.5, 'pos': 0.5, 'compound': 0.4585}


The output is 50% positive ad 50% neutral. The compound score is 0.4585.

4\. **Calculate the compound sentiment scores of the first 1,000 training data. Convert the final scores to 0 (negative) and 1 (positive).**

In [8]:
train_examples_batch, train_labels_batch = next(iter(train_data.batch(1000)))

In [9]:
score = [0 for x in range(1000)]
for i in range(1000):
    text = train_examples_batch.numpy()[i].decode("utf-8")
    sent = analyzer.polarity_scores(text)['compound']
    if(sent > 0):
        score[i] = 1

5\. **Evaluate the performance of the predicted sentiment socres using the `classification_report` function. How do you analyze your results?**

In [10]:
print(metrics.classification_report(train_labels_batch, score, target_names=['negative', 'positive']))

              precision    recall  f1-score   support

    negative       0.78      0.53      0.63       490
    positive       0.66      0.85      0.74       510

    accuracy                           0.70      1000
   macro avg       0.72      0.69      0.69      1000
weighted avg       0.71      0.70      0.69      1000



# Deep learning-based sentiment analysis

In this part of the practical, we are going to use pre-trained word embedding models from TensorFlow Hub (https://tfhub.dev/) to do sentiment classification on movie reviews. TensorFlow Hub is a repository of trained machine learning models.

6\. **Use a pre-trained embedding model from TensorFlow Hub called `"google/nnlm-en-dim50/2"`, and create a Keras embedding layer that uses this model to embed the sentences, and try it out on a couple of input examples.**

In [11]:
# Token based text embedding trained on English Google News 7B corpus.
tf_keras.utils.set_random_seed(42)

embedding_50_url = "https://tfhub.dev/google/nnlm-en-dim50/2"

embedding_layer_50 = hub.KerasLayer(
    embedding_50_url,
    input_shape=[],
    dtype=tf.string,
    trainable=True,
    name="embedding_layer",
)

embedding_layer_50(train_examples_batch[:3])

<tf.Tensor: shape=(3, 50), dtype=float32, numpy=
array([[ 0.5423195 , -0.0119017 ,  0.06337538,  0.06862972, -0.16776837,
        -0.10581174,  0.16865303, -0.04998824, -0.31148055,  0.07910346,
         0.15442263,  0.01488662,  0.03930153,  0.19772711, -0.12215476,
        -0.04120981, -0.2704109 , -0.21922152,  0.26517662, -0.80739075,
         0.25833532, -0.3100421 ,  0.28683215,  0.1943387 , -0.29036492,
         0.03862849, -0.7844411 , -0.0479324 ,  0.4110299 , -0.36388892,
        -0.58034706,  0.30269456,  0.3630897 , -0.15227164, -0.44391504,
         0.19462997,  0.19528408,  0.05666234,  0.2890704 , -0.28468323,
        -0.00531206,  0.0571938 , -0.3201318 , -0.04418665, -0.08550783,
        -0.55847436, -0.23336391, -0.20782952, -0.03543064, -0.17533456],
       [ 0.56338924, -0.12339553, -0.10862679,  0.7753425 , -0.07667089,
        -0.15752277,  0.01872335, -0.08169781, -0.3521876 ,  0.4637341 ,
        -0.08492756,  0.07166859, -0.00670817,  0.12686075, -0.19326553,
 

Here you see that no matter the length of the input text, the output shape of the embeddings is: (`num_examples`, `embedding_dimension`).

7\. **Build a deep learning model using the embedding layer and one hidden layer.**

In [12]:
model_50 = tf_keras.Sequential()
model_50.add(embedding_layer_50)
model_50.add(tf_keras.layers.Dense(16, activation='relu', name="hidden_layer"))
model_50.add(tf_keras.layers.Dense(1, name="sentiment_probability/output_layer"))

model_50.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_layer (KerasLaye  (None, 50)                48190600  
 r)                                                              
                                                                 
 hidden_layer (Dense)        (None, 16)                816       
                                                                 
 sentiment_probability/outp  (None, 1)                 17        
 ut_layer (Dense)                                                
                                                                 
Total params: 48191433 (183.84 MB)
Trainable params: 48191433 (183.84 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


8\. **Compile and train the model for 10 epochs in batches of 512 samples.**

In [13]:
model_50.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

In [14]:
start_50 = time.perf_counter()

history_50 = model_50.fit(train_data.shuffle(10000).batch(512),
                    epochs=10,
                    validation_data=validation_data.batch(512),
                    verbose=1)

training_seconds_50 = time.perf_counter() - start_50
print(f"Training time: {training_seconds_50:.1f} seconds")

Epoch 1/10
30/30 [==============================] - 50s 2s/step - loss: 0.9507 - accuracy: 0.5561 - val_loss: 0.7713 - val_accuracy: 0.6133
Epoch 2/10
30/30 [==============================] - 43s 1s/step - loss: 0.5946 - accuracy: 0.7081 - val_loss: 0.5927 - val_accuracy: 0.7250
Epoch 3/10
30/30 [==============================] - 60s 2s/step - loss: 0.4316 - accuracy: 0.8171 - val_loss: 0.5043 - val_accuracy: 0.7891
Epoch 4/10
30/30 [==============================] - 44s 1s/step - loss: 0.3206 - accuracy: 0.8780 - val_loss: 0.4964 - val_accuracy: 0.8210
Epoch 5/10
30/30 [==============================] - 44s 1s/step - loss: 0.2379 - accuracy: 0.9156 - val_loss: 0.5053 - val_accuracy: 0.8425
Epoch 6/10
30/30 [==============================] - 43s 1s/step - loss: 0.1802 - accuracy: 0.9406 - val_loss: 0.5274 - val_accuracy: 0.8488
Epoch 7/10
30/30 [==============================] - 44s 1s/step - loss: 0.1390 - accuracy: 0.9572 - val_loss: 0.5431 - val_accuracy: 0.8571
Epoch 8/10
30/30 [==

9\. **Evaluate the model on the test set.**

In [15]:
results_50 = model_50.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model_50.metrics_names, results_50):
  print("%s: %.3f" % (name, value))

49/49 - 10s - loss: 0.7168 - accuracy: 0.8443 - 10s/epoch - 201ms/step
loss: 0.717
accuracy: 0.844


This fairly simple approach achieves an accuracy of about 85%.

In [16]:
# we can use a trained model for prediction, by calling .predict()
model_50_probability = float(
    model_50.predict(
        tf.constant([vader_example_sentence]), # you can replace vader_example_sentence with any other string
        verbose=0,
    )[0, 0]
)

model_50_label = "positive" if model_50_probability >= 0.5 else "negative"

print("Sentence:", vader_example_sentence)
print(f"Predicted positive probability: {model_50_probability:.4f}")
print("Predicted label:", model_50_label)

Sentence: you cannot be negative
Predicted positive probability: 0.5171
Predicted label: positive


10\. **For your next experiment load a more complex pretrained word embedding for the embedding layer. Train and evaluate your model.**

In [18]:
tf_keras.utils.set_random_seed(42)

embedding_128_url = "https://tfhub.dev/google/nnlm-en-dim128-with-normalization/2"
embedding_layer_128 = hub.KerasLayer(embedding_128_url, input_shape=[],
                                     dtype=tf.string, trainable=True, name="embedding_layer")
# embedding_layer_128(train_examples_batch[:3])

Here we tried google/nnlm-en-dim128-with-normalization/2 - trained with the same NNLM (Neural Network Language Model) architecture on the same data as google/nnlm-en-dim50/2, but with a larger embedding dimension. Larger dimensional embeddings can improve on your task but it may take longer to train your model. This new model has additional text normalization such as removing punctuation. This can help if the text in your task contains additional characters or punctuation. You can try more pretrained embeddings from [TensorFlow Hub](https://tfhub.dev/s), for example BERT, but rememeber that these are huge models and need a lot of training time.

In Practical 10, you will fine-tune and fit BERT!

In [19]:
model_128 = tf_keras.Sequential()
model_128.add(embedding_layer_128)
model_128.add(tf_keras.layers.Dense(16, activation='relu', name="hidden_layer"))
model_128.add(tf_keras.layers.Dense(1, name="sentiment_probability/output_layer"))

model_128.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding_layer (KerasLaye  (None, 128)               124642688 
 r)                                                              
                                                                 
 hidden_layer (Dense)        (None, 16)                2064      
                                                                 
 sentiment_probability/outp  (None, 1)                 17        
 ut_layer (Dense)                                                
                                                                 
Total params: 124644769 (475.48 MB)
Trainable params: 124644769 (475.48 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [20]:
model_128.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

In [21]:
start_128 = time.perf_counter()

history_128 = model_128.fit(train_data.shuffle(10000).batch(512),
                    epochs=10,
                    validation_data=validation_data.batch(512),
                    verbose=1)

training_seconds_128 = time.perf_counter() - start_128
print(f"Training time: {training_seconds_128:.1f} seconds")

Epoch 1/10
30/30 [==============================] - 110s 4s/step - loss: 0.8169 - accuracy: 0.6461 - val_loss: 0.5443 - val_accuracy: 0.7349
Epoch 2/10
30/30 [==============================] - 107s 4s/step - loss: 0.3672 - accuracy: 0.8590 - val_loss: 0.4341 - val_accuracy: 0.8482
Epoch 3/10
30/30 [==============================] - 143s 5s/step - loss: 0.2326 - accuracy: 0.9253 - val_loss: 0.3994 - val_accuracy: 0.8675
Epoch 4/10
30/30 [==============================] - 125s 4s/step - loss: 0.1573 - accuracy: 0.9569 - val_loss: 0.4156 - val_accuracy: 0.8818
Epoch 5/10
30/30 [==============================] - 116s 4s/step - loss: 0.1094 - accuracy: 0.9749 - val_loss: 0.4193 - val_accuracy: 0.8845
Epoch 6/10
30/30 [==============================] - 128s 4s/step - loss: 0.0802 - accuracy: 0.9855 - val_loss: 0.4348 - val_accuracy: 0.8850
Epoch 7/10
30/30 [==============================] - 115s 4s/step - loss: 0.0619 - accuracy: 0.9906 - val_loss: 0.4629 - val_accuracy: 0.8862
Epoch 8/10
30

In [22]:
results_128 = model_128.evaluate(test_data.batch(512), verbose=2)

for name, value in zip(model_128.metrics_names, results_128):
  print("%s: %.3f" % (name, value))

49/49 - 23s - loss: 0.5998 - accuracy: 0.8630 - 23s/epoch - 467ms/step
loss: 0.600
accuracy: 0.863


In [25]:
# we can use a trained model for prediction, by calling .predict()
model_128_probability = float(
    model_128.predict(
        tf.constant([vader_example_sentence]), # you can replace vader_example_sentence with any other stringnce
        verbose=0,
    )[0, 0]
)

model_128_label = "positive" if model_128_probability >= 0.5 else "negative"

print("Sentence:", vader_example_sentence)
print(f"Predicted positive probability: {model_128_probability:.4f}")
print("Predicted label:", model_128_label)

Sentence: you cannot be negative
Predicted positive probability: 0.3016
Predicted label: negative


Let's compare two models.

In [24]:
comparison = pd.DataFrame({
    "model": ["NNLM 50", "NNLM 128 normalized"],
    "parameters": [model_50.count_params(), model_128.count_params()],
    "training_seconds": [training_seconds_50, training_seconds_128],
    "test_accuracy": [results_50[1], results_128[1]],
    "test_loss": [results_50[0], results_128[0]]
})

comparison

,model,parameters,training_seconds,test_accuracy,test_loss
0,NNLM 50,48191433,558.237901,0.84428,0.716781
1,NNLM 128 normalized,124644769,1378.757536,0.86304,0.599753
